janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphStep.java

```diff
@@ -293,5 +307,39 @@ public int hashCode() {
         result = 31 * result + (orders != null ? orders.hashCode() : 0);
         return result;
     }
+
+    private final class ProfiledIterator<E extends JanusGraphElement> extends CloseableAbstractIterator<E> {
+        private final QueryProfiler profiler;
+        private Iterator<E> iterator;
+        private boolean timerRunning;
+
+        private ProfiledIterator(QueryProfiler profiler) {
+            this.profiler = profiler;
+            profiler.startTimer();
```

Comment:
Is there a reason the timer should start once the iterator is created? In your case, the measured time frame includes the `builder.iterables(...).iterator()` call in line 179. No matter if the time spent there should or should not be recorded, there is currently no mechanism in place to enforce it either way. Currently, everything that happens between the iterator construction and its use is included in the measurement, which does not intuitively make sense to me.

Reply:
Agree that it does not make much intuitive sense. The reason is, a couple of underlying iterators, e.g. SubqueryIterator, eagerly fetches results in their constructors. Thus, the backend query is already executed before the next() or hasNext() is called the first time.

Reply:
I see. There are two ideas that come to my mind:

1. Start the timer for the `ProfiledIterator` explicitly by calling a method. This indicates more clearly that there is a reason to create the `ProfiledIterator` instance this early.
2. Delay the construction of the underlying iterator, e.g. by passing a `Supplier<Iterator<E>>` as a parameter to `setIterator()` (or the constructor). The supplier can then be evaluated on the first call of `next()` which is also when the timer will be started.

The second solution has the advantage that the timer is automatically started once the iterator is actually accessed for the first time. However, that would make calls of `setIterator()` look a little bit more complex:
```java
iterator.setIterator(() -> (Iterator<E>) builder.iterables(query, graphClass).iterator());
```

But of course, we can also move that to the constructor then:
```java
ProfiledIterator iterator = new ProfiledIterator(profiler, () -> (Iterator<E>) builder.iterables(query, graphClass).iterator());
```